src/xtgeo/grid3d/grid.py

```diff
@@ -1984,6 +1984,56 @@ def get_bulk_volume(
             self, name=name, precision=precision, asmasked=asmasked
         )
 
+    def get_phase_volume(
+        self,
+        woc: float | GridProperty,
+        goc: float | GridProperty,
+        boundary: Polygons | None = None,
+        asmasked: bool = True,
+        precision: Literal[1, 2, 4] = 2,
+    ) -> tuple[GridProperty, GridProperty, GridProperty]:
+        """Return the geometric phase cell volume for all cells as three GridProperty
+        object namely gas_bulkvol, oil_bulkvol, water_bulkvol.
+
+        The volume calculation is constrained to area within boundary polygon.
+
+        A bulk volume of a cornerpoint cell is actually a non-trivial and a non-unique
+        entity. The volume is approximated by dividing the cell (hexahedron) into
+        6 tetrehedrons; there is however a large number of ways to do this division.
+
+        As default (precision=2) an average of two different ways to divide the cell
+        into tetrahedrons is averaged.
+
+        Args:
+            woc: Water oil contact
+            goc: Gas oil contact
+            boundary: Boundary area
```

Comment:
Need to say for boundary that if the object contains several polygons, only the first one applied